dhis2/dhis2-core
# feat: OpenAPI default values for parameters [DHIS2-15893]

### Summary
Adds the `default` information to the OpenAPI spec. The `default` is a field of the `schema` of a parameter object. 

It is supported for simple types only as the spec demands that the value given is of the same type as the schema's value `type`. In theory we could support arrays of simple types and similar things but the complexity of adjusting and converting between a Java string and the appropriate JSON value for the OpenAPI spec is not worth the benefit. 

### Usage in Source
The default value can be specified on endpoint method parameters in 2 ways (in order of precedence):

* `@OpenApi.Param(defaultValue = "x")`
* `@RequestParam(defaultValue = "x")`

For parameter objects the default value can be specified on 3 ways (in order of precedence):
* `@OpenApi.Property(defaultValue = "DATA_VALUE")` on the field
* `@JsonProperty(defaultValue = "DATA_VALUE")` on the field
* The field's initial value, e.g. `private FileResourceDomain domain = FileResourceDomain.DATA_VALUE` 

### OpenAPI API Changes
A new flag `failOnInconsistency` was added to all OpenAPI endpoints. If set true, a inconsistency between parameter's `required` and its `default` will cause an exception, otherwise it will just log a warning. Inconsistency here means a parameter was marked as required while also having a default value. The added flag is intended to not just flip the behaviour of this new validation but any similar validation of inconsistency of declaration that we might add in the future. Such validations are evaluated in the `ApiFinalise` phase after the `Api` model has been build in the analysis phase. 

### Automatic Testing
A new test scenario was added that checks a couple of known defaults are present in the spec as declared in the source.

### Manual Testing
* get a OpenAPI spec for one of our APIs, e.g. https://play.dhis2.org/dev/api/users/openapi.json
* search the document for `GistParams.pageSize` definition
* check the `schema` object has a `default` property of `50` 
* alternatively search for `"default":` in the document to see the different defaults defined now
* compare this to earlier versions where such `"default":` properties are not included in the `scheme` object of parameters
